services/image-rec-master/service/tasks.py

```diff
@@ -0,0 +1,383 @@
+of the image recognition pipelines, with auto and manual image rec
+do the "middle parts".
+
+There are two distict pipelines at play: images and targets.
+
+This module contains tasks for the queueing / re-queueing of images,
+and the submission and removal of targets.
+
+Image pipeline:
+    New images are checked for on an interval and placed into
+    'all-images'. New Images that do not need to be checked are
+    placed into 'skipped-auto' and 'skipped-manual' for auto and
+    manual image rec. Otherwise, new images are placed into
+    'unprocessed-auto' and 'unprocessed-manual'.
+
+    Auto workers take images from 'unprocessed-auto' and place them
+    in 'processing-auto'. Upon successful processing, they are moved
+    into 'processed-auto' (and at the same time a target is placed
+    in the target pipeline if necessary). If the auto worker errors
+    during processing, the image will remain in 'processing-auto'. A
+    task watches 'processing-auto' to determine if images have
+    stalled. If they have, they are placed into 'retrying-auto' and
+    are also placed in 'unprocessed-auto' if they are not already in
+    'retrying-auto' or 'errored-auto'. Otherwise, if they are already
+    in 'retrying-auto' (thus have failed processing before), they are
+    placed into 'errored-auto' and removed from 'retrying-auto'.
+
+    Manual workers take images from 'unprocessed-manual' and move
+    them into 'processed-manual' directly. Manual workers may also
+    move images from 'skipped-manual' to 'processed-manual'.
+
+Target pipeline:
+    Unique targets are placed into 'all-targets' and
+    'unsubmitted-targets'. Uniqueness is checked by auto workers by
+    viewing all targets submitted at that point while uniqueness is
+    not checked for manual targets.
+
+    Targets from 'unsubmitted-targets' are moved into
+    'submitting-targets' to indicate that they are being submitted.
+    Upon successful submission, they are moved into
+    'submitted-targets'. Otherwise, on submission failure, they are
+    placed into 'errored-targets'. Submission failures are not due
+    to API unavailability, but instead invalid submission data
+    (`400`s).
+
+    Targets may also be removed. Removal requests are placed into
+    'unremoved-targets', i.e. they are not removed yet. Targets from
+    'unremoved-targets' are moved into 'removing-targets' to indicate
+    that they are being removed. If the target is in
+    'unsubmitted-targets', it is moved directly into
+    'removed-targets'. If the target does not exist on the interop
+    server, it is ignored. Otherwise, upon successful removal, it is
+    placed into 'removed-targets'. Removed targets that were
+    submitted remain in 'submitted-targets'. There is no error
+    condition for removal.
+
+Note that are both Redis lists and sets involved. Lists are used
+for items that worker tasks pop elements directly off of, while sets
+are used for reading on demand.
+"""
```

Comment:
We should pull out a fancy ascii diagram

Reply:
I tried but wasn't feeling creative enough lol.